### PR TITLE
Add a script for benchmarking performance for Ray developers.

### PR DESCRIPTION
**Pull request by robertnishihara**
## Why are these changes needed?

Big PRs sometimes introduce performance regressions, and we don't have a reliable way to prevent these performance regressions. This PR is a first step toward helping avoid performance regressions by providing a script that can be run to benchmark performance.

Before merging large PRs, the author should run this script on the PR branch as well as on the code prior to the PR and report the two results. Care should be taken to make sure the results are not noisy (e.g., this may mean running the test on a machine in EC2 that isn't doing anything else).

There's still a lot more that will need to be done in the future, e.g., ideally we'd have a single script that takes in a PR number, launches a cluster with the autoscaler, installs Ray from that PR, runs the performance tests, then does the same for the previous release or perhaps the current master, and then evaluates the difference in performance.

We've historically had trouble reducing the variance of our performance measurements, so doing a good job with that will be critical for making these tests useful.

NOTE: This PR also removes a bunch of older scripts related to performance testing, which are no longer in use.

### CODE REVIEW

File: ci/performance_tests/test_performance.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    default=False,
+    action="store_true",
+    help="True if the object store should not be warmed up. This could cause "
+    "the benchmarks to appear slower than usual.")
+parser.add_argument(
+    "--warmup-object-store",
+    required=False,
+    type=bool,
+    default=True,
+    help="The number of nodes to simulate in the cluster.")
```
Comment (pcmoritz) [inline ci/performance_tests/test_performance.py]: The help message doesn't correspond to the flag

File: ci/performance_tests/test_performance.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    type=int,
+    help="The number of nodes to simulate in the cluster.")
+parser.add_argument(
+    "--skip-object-store-warmup",
+    default=False,
+    action="store_true",
+    help="True if the object store should not be warmed up. This could cause "
+    "the benchmarks to appear slower than usual.")
+parser.add_argument(
+    "--warmup-object-store",
```
Comment (pcmoritz) [inline ci/performance_tests/test_performance.py]: We don't need both `warmup-object-store` and `skip-object-store-warmup`, right?

File: ci/performance_tests/test_performance.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    default=True,
+    help="The number of nodes to simulate in the cluster.")
+parser.add_argument(
+    "--redis-address",
+    required=False,
+    type=str,
+    help="The address of the cluster to connect to. If this is ommitted, then "
+    "a cluster will be started locally (on a single machine).")
+
+num_redis_shards = 2
```
Comment (pcmoritz) [inline ci/performance_tests/test_performance.py]: Having global variables with snake_case is unusual (normally they would be ALL_CAPS), how about make them arguments in the argument parser with the right defaults?

File: ci/performance_tests/test_performance.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+            resources={str(i): 500},
+            object_store_memory=object_store_memory,
+            redis_max_memory=redis_max_memory)
+    ray.init(redis_address=cluster.redis_address)
+
+    return cluster
+
+
+def wait_for_and_check_cluster_configuration(num_nodes):
+    """Check that the cluster's custom resources are properly configured.
```
Comment (pcmoritz) [inline ci/performance_tests/test_performance.py]: This should use the usual docstring format.

File: ci/performance_tests/test_performance.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+             for node in ray.nodes()] == [{str(i), "CPU"} for i in range(3)]):
+        raise RuntimeError(
+            "The ith node in the cluster should have a "
+            "custom resource called 'i' with quantity "
+            "500. The nodes are\n%s", ray.nodes())
+    if not ([[
+            resource_quantity
+            for resource_name, resource_quantity in node["Resources"].items()
+            if resource_name != "CPU"
+    ] for node in ray.nodes()] == 3 * [[500.0]]):
```
Comment (pcmoritz) [inline ci/performance_tests/test_performance.py]: the `3` should be `num_nodes`

File: ci/performance_tests/test_performance.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+    durations = run_multiple_trials(one_thousand_serial_tasks_remote_node, 10)
+    logger.warning(
+        "one_thousand_serial_tasks_remote_node \n"
+        "    min:  %.2gs\n"
+        "    mean: %.2gs\n"
+        "    std:  %.2gs", np.min(durations), np.mean(durations),
+        np.std(durations))
+
+    def ten_thousand_parallel_tasks_local():
```
Comment (pcmoritz) [inline ci/performance_tests/test_performance.py]: Why is this called "parallel tasks"? Looks serial to me.

**Review by pcmoritz — APPROVED** [review]

File: ci/performance_tests/test_performance.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    durations = []
+    for _ in range(num_trials):
+        start = time.time()
+        f()
+        durations.append(time.time() - start)
+    return durations
+
+
+def test_tasks(num_nodes):
+    def one_thousand_serial_tasks_local_node():
```
Comment (raulchen) [inline ci/performance_tests/test_performance.py]: we can also add a test that only submit tasks without getting results.

**Pull request merged**